pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java

```diff
@@ -324,32 +342,41 @@ protected void internalGrantPermissionsOnTopic(final AsyncResponse asyncResponse
-                                      actions));
-                          }
-                      }
-                      return future.thenCompose(unused -> grantPermissionsAsync(topicName, role, actions))
-                              .thenAccept(unused -> asyncResponse.resume(Response.noContent().build()));
-                  }))).exceptionally(ex -> {
+                        grantPermissionsAsync(TopicName.get(topicName.getPartitionedTopicName()), role, actions)
+                                .thenAccept(unused -> asyncResponse.resume(Response.noContent().build()))))
+                .exceptionally(ex -> {
                     Throwable realCause = FutureUtil.unwrapCompletionException(ex);
                     log.error("[{}] Failed to get permissions for topic {}", clientAppId(), topicName, realCause);
                     resumeAsyncResponseExceptionally(asyncResponse, realCause);
                     return null;
                 });
     }
 
-    private CompletableFuture<Void> revokePermissionsAsync(String topicUri, String role) {
+    private CompletableFuture<Void> revokePermissionsAsync(TopicName topicName, String role, int numPartitions) {
+        String topicUri = topicName.toString();
         return namespaceResources().getPoliciesAsync(namespaceName).thenCompose(
                 policiesOptional -> {
                     Policies policies = policiesOptional.orElseThrow(() ->
                             new RestException(Status.NOT_FOUND, "Namespace does not exist"));
+                    CompletableFuture<Void> future = CompletableFuture.completedFuture(null);
+                    // do compatible with previous pulsar version
+                    // revoke all the partition permissions granted in previous version
+                    future = future.thenComposeAsync(unused ->
+                        namespaceResources().setPoliciesAsync(namespaceName, p -> {
+                            if (numPartitions > 0) {
+                                for (int i = 0; i < numPartitions; i++) {
+                                    p.auth_policies.getTopicAuthentication().computeIfPresent(
+                                            topicName.getPartition(i).toString(), (k, roles) -> {
+                                                roles.remove(role);
+                                                if (roles.isEmpty()) {
+                                                    return null;
+                                                }
+                                                return roles;
+                                            });
+                                }
+                            }
+                            return p;
+                        }));
                     if (!policies.auth_policies.getTopicAuthentication().containsKey(topicUri)
                             || !policies.auth_policies.getTopicAuthentication().get(topicUri).containsKey(role)) {
```

Comment:
I wonder if instead of failing here, we should just attempt to remove permission on every partition, even if it isn't in the map? I think that would be the expected behavior. This would also be backwards compatible.

Reply:
Fixed by #17393.